python/lsst/daf/butler/core/connectionString.py

```diff
@@ -0,0 +1,168 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("ConnectionString", "ConnectionStringFactory")
+
+from sqlalchemy.exc import NoSuchModuleError
+
+from sqlalchemy.engine import url
+from lsst.daf.butler.core.dbAuth import DbAuth
+from lsst.daf.butler.core.registry import RegistryConfig
+
+DB_AUTH_ENVVAR = "LSST_DB_AUTH"
+"""Default name of the environmental variable that will be used to locate DB
+credentials configuration file. """
+
+DB_AUTH_PATH = "~/.lsst/db-auth.yaml"
+"""Default path at which it is expected that DB credentials are found."""
+
+
+class ConnectionString(url.URL):
+    """Builds a connection string of the format:
+
+        dialect+driver://user:password@host:port/database
+
+    by parsing (in order of precedence):
+
+    1. 'db' key in RegistryConfig,
+    2. RegistryConfig for explicitly defined keys
+    3. and DbAuth instance with which it was created.
+
+    The 'db' string must be a connection URL, i.e. must start with a database
+    dialect and contain, at least, the `database` component or the host name.
+
+    Components `username`, `password`, `host`, `port` or `database` can also be
+    specified explicitly as keys in the config when not part of the `db` string
+    in Config.
+    When `username` and `password` are not part of the `db` string or stated
+    explicitly in Config they will be read from `~/.lsst/db-auth.paf` file by
+    matching matching `host` and `port` components of provided URL.
+
+    Parameters
+    ----------
+    dbAuth : `DbAuth`
+        DbAuth instance to use in order to resolve username and passwords.
+    conUrl : `sqlalchemy.engine.URL`
+        URL instance from which connection string will be created
+    """
+    def __init__(self, dbAuth, conUrl):
+        super().__init__(conUrl.drivername, **conUrl.translate_connect_args())
+        self._dbAuth = dbAuth
+
+    @property
+    def dialect(self):
+        """Returns the dialect part of the connection string."""
+        return self.get_backend_name()
+
+    @property
+    def driver(self):
+        """Returns the driver part of the connection string."""
+        return self.get_driver_name()
+
+    def checkDialectDriverExist(self):
+        """Verifies that the targeted dialect and driver exist and are
+        installed.
+
+        Attempts to load the correct dialect and driver from SQLAlchemy. If
+        load is successfull, returns True, otherwise returns False.
+        """
+        try:
+            self._conStr.get_dialect()
+        except NoSuchModuleError:
+            return False
+        return True
+
+    def asDict(self):
+        """Returns connection string arguments as a dictionary."""
+        return self._conStr.translate_connect_args()
+
+
+class ConnectionStringFactory:
+    """Factory for `ConnectionString` instances.
+
+    Parameters
+    ----------
+    path : `str` or None, optional
```

Comment:
I don't think you need this generality.  These parameters were meant to allow the application to configure DbAuth.  The application itself doesn't need to be configured.  You can remove all these parameters and just use the module defaults.

Reply:
Application now not configurable. Tests changed to reflect this - the module `DB_AUTH_PATH` and `DB_AUTH_ENVVAR` vars are now being retargeted.